hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/PipelineChoosePolicy.java

```diff
@@ -30,8 +30,18 @@ public interface PipelineChoosePolicy {
    * Given an initial list of pipelines, return one of the pipelines.
    *
    * @param pipelineList list of pipelines.
-   * @return one of the pipelines.
+   * @return one of the pipelines or null if no pipeline can be selected.
    */
   Pipeline choosePipeline(List<Pipeline> pipelineList,
       PipelineRequestInformation pri);
+
+  /**
+   * Given a list of pipelines, return the index of the chosen pipeline.
+   * @param pipelineList List of pipelines
+   * @param pri          PipelineRequestInformation
+   * @return Index in the list of the chosen pipeline, or -1 if no pipeline
+   *         could be selected.
+   */
+  int choosePipelineIndex(List<Pipeline> pipelineList,
+      PipelineRequestInformation pri);
```

Comment:
SCM can be configured with custom `PipelineChoosePolicy`, so we have to assume there may be implementations "in the wild".  In order to avoid breaking them, this should be added as a `default` method in the interface.  Considering `HealthyPipelineChoosePolicy` throws exception, such default implementation may not be feasible.

Reply:
I find it very unlikely that anyone has an implementation of the pipeline choosing policy. Whats the alternative to never being able to change this?

The healthy policy isn't good IMO, as it mutates the array it is passed. That could have some strange side effect, if for example the caller was using an iterator over the array or something.

I wonder if we could fix he Healthy policy somehow to avoid making it throw on the exception.

Reply:
How about something like this as the default implementation?

```java
  default int choosePipelineIndex(List<Pipeline> pipelineList,
      PipelineRequestInformation pri) {
    return pipelineList == null || pipelineList.isEmpty() ? -1 : 0;
  }
```

i.e. return the first pipeline, unless there is none.

We can probably state in the release notes that custom implementations need to override it to ensure that the two methods are consistent.